codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java

```diff
@@ -132,4 +143,43 @@ protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> ou
         }
         ReferenceCountUtil.retain(obj);
     }
+
+    private Http2Headers toHttp2Headers(final HttpMessage msg) {
+        if (msg instanceof HttpRequest) {
+            msg.headers().set(
+                    HttpConversionUtil.ExtensionHeaderNames.SCHEME.text(),
+                    scheme.name());
+        }
+
+        return HttpConversionUtil.toHttp2Headers(msg, validateHeaders);
+    }
+
+    private HttpMessage newMessage(final int id,
+                                   final Http2Headers headers) throws Http2Exception {
+        return isServer ?
+                HttpConversionUtil.toHttpRequest(id, headers, validateHeaders) :
+                HttpConversionUtil.toHttpResponse(id, headers, validateHeaders);
+    }
+
+    private FullHttpMessage newFullMessage(final int id,
+                                           final Http2Headers headers,
+                                           final ByteBufAllocator alloc) throws Http2Exception {
+        return isServer ?
+                HttpConversionUtil.toFullHttpRequest(id, headers, alloc, validateHeaders) :
+                HttpConversionUtil.toFullHttpResponse(id, headers, alloc, validateHeaders);
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        super.handlerAdded(ctx);
+
+        scheme = isSsl(ctx) ? HttpScheme.HTTPS : HttpScheme.HTTP;
+    }
+
+    protected boolean isSsl(final ChannelHandlerContext ctx) {
+        final Channel ch = ctx.channel();
+        final Channel connChannel = (ch instanceof Http2StreamChannel) ? ch.parent() : ch;
+        final ChannelPipeline p = connChannel.pipeline();
+        return null != p.get(SslHandler.class) || null != p.get(SniHandler.class);
```

Comment:
if SNI is in the pipeline we should wait until a `SslHandshakeCompletionEvent` is received in the pipeline before we consider the connection to be `HTTPS`.

For example:
```java
@Override
public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
  if (evt instanceof SslHandshakeCompletionEvent) {
    // SslHandler should be in the pipeline
  }
}
```

Reply:
+1

Reply:
@lionelli1 @Scottmitch that said we need to ensure we do this as part of the "correct" pipeline which may be the parent pipeline or not depending on if `Http2MultiplexCodex` is used or not.

Reply:
but for most of the use cases where this handler is added at the child stream, the SslHandshakeCompletionEvent would've happened at the parent channel by the time it is added to the stream channels. the new child channels won't be notified of this event upon creation, would it?

Reply:
yep ... maybe its ok for now to just do this only if the its not a "child channel" for now

Reply:
ok changed to checking on SslHandler with the assumption that this handler will mostly be used on an `Http2StreamChannel`